File: lib/geocode.ts
/**
 * Reverse geocode coordinates using Nominatim (OpenStreetMap).
 * Returns a normalized display address string or null if lookup fails.
 */
export async function reverseGeocodeInstitution(
	lat: number,
	lon: number,
): Promise<{ addressLine: string; city?: string; state?: string } | null> {
	try {
		const controller = new AbortController();
		const timeoutId = setTimeout(() => controller.abort(), 8000);

		const res = await fetch(
			`https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=${lat}&lon=${lon}`,
			{
				headers: { "User-Agent": "sedekahje-bot" },
				signal: controller.signal,
			},
		);

		clearTimeout(timeoutId);

		if (!res.ok) return null;

		const data = (await res.json()) as {
			address?: {
				road?: string;
				house_number?: string;
				suburb?: string;
				village?: string;
				town?: string;
				city?: string;
				municipality?: string;
				state?: string;
				postcode?: string;
				country?: string;
			};
			display_name?: string;
		};

		const addr = data.address;
		if (!addr) return null;

		const parts: string[] = [];
		const street = [addr.house_number, addr.road].filter(Boolean).join(" ");
		if (street) parts.push(street);
		const sub = addr.suburb ?? addr.village ?? addr.town ?? addr.municipality;
		if (sub) parts.push(sub);
		if (addr.city && sub !== addr.city) parts.push(addr.city);
		if (addr.state) parts.push(addr.state);
		if (addr.postcode) parts.push(addr.postcode);
		if (addr.country) parts.push(addr.country);

		const addressLine =
			parts.length > 0 ? parts.join(", ") : (data.display_name ?? "");
		if (!addressLine) return null;

		return {
			addressLine,
			city: addr.city ?? addr.town ?? addr.village ?? addr.municipality,
			state: addr.state,
		};
	} catch (err) {
		console.error("[reverse geocode]", err);
		return null;
	}
}

/**
 * Geocode an institution using Nominatim (OpenStreetMap).
 * Returns [latitude, longitude] or null if geocoding fails.
 */
export async function geocodeInstitution(
	name: string,
	city: string,
	state: string,
): Promise<[number, number] | null> {
	try {
		const query = `${name}, ${city}, ${state}, Malaysia`;
		const controller = new AbortController();
		const timeoutId = setTimeout(() => controller.abort(), 5000);

		const res = await fetch(
			`https://nominatim.openstreetmap.org/search?format=jsonv2&limit=1&q=${encodeURIComponent(query)}`,
			{
				headers: { "User-Agent": "sedekahje-bot" },
				signal: controller.signal,
			},
		);

		clearTimeout(timeoutId);

		if (!res.ok) return null;

		const results = (await res.json()) as Array<{ lat: string; lon: string }>;
		if (results.length === 0) return null;

		const lat = Number.parseFloat(results[0].lat);
		const lon = Number.parseFloat(results[0].lon);
		if (Number.isNaN(lat) || Number.isNaN(lon)) return null;

		return [lat, lon];
	} catch {
		return null;
	}
}

/**
 * Geocode an institution using Google Geocoding API.
 * Returns [latitude, longitude] or null if geocoding fails.
 * Requires GOOGLE_GEOCODING_API_KEY env var.
 */
export async function geocodeWithGoogle(
	name: string,
	city: string,
	state: string,
): Promise<[number, number] | null> {
	const apiKey = process.env.GOOGLE_GEOCODING_API_KEY;
	if (!apiKey) return null;

	try {
		const query = `${name}, ${city}, ${state}, Malaysia`;
		const url = new URL("https://maps.googleapis.com/maps/api/geocode/json");
		url.searchParams.set("address", query);
		url.searchParams.set("key", apiKey);
		url.searchParams.set("region", "my");
		url.searchParams.set("language", "ms");

		const controller = new AbortController();
		const timeoutId = setTimeout(() => controller.abort(), 8000);

		const res = await fetch(url.toString(), {
			headers: { Accept: "application/json" },
			signal: controller.signal,
		});

		clearTimeout(timeoutId);

		if (!res.ok) return null;

		const data = (await res.json()) as {
			status: string;
			results?: Array<{
				geometry?: { location?: { lat?: number; lng?: number } };
			}>;
		};

		if (data.status !== "OK" || !data.results?.length) return null;

		const loc = data.results[0].geometry?.location;
		if (typeof loc?.lat !== "number" || typeof loc?.lng !== "number")
			return null;

		return [loc.lat, loc.lng];
	} catch {
		return null;
	}
}

/**
 * Geocode with Google Maps first, fallback to Nominatim.
 */
export async function geocodeInstitutionWithFallback(
	name: string,
	city: string,
	state: string,
): Promise<[number, number] | null> {
	const google = await geocodeWithGoogle(name, city, state);
	if (google) return google;

	return geocodeInstitution(name, city, state);
}

File: src/geocode.rs
use crate::config::{app_url, Config};

/// Geocode an institution name/city/state into [lat, lng] using Google first,
/// then Nominatim as fallback. Returns None on failure.
pub async fn geocode_institution_with_fallback(
    cfg: &Config,
    name: &str,
    city: &str,
    state: &str,
) -> Option<(f64, f64)> {
    if let Some(key) = cfg.google_geocoding_api_key.as_ref() {
        if !key.is_empty() {
            if let Some(c) = geocode_google(cfg, name, city, state).await {
                return Some(c);
            }
        }
    }
    geocode_nominatim(cfg, name, city, state).await
}

async fn geocode_google(
    cfg: &Config,
    name: &str,
    city: &str,
    state: &str,
) -> Option<(f64, f64)> {
    let key = cfg.google_geocoding_api_key.clone()?;
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .ok()?;
    let query = format!("{name}, {city}, {state}");
    let url = format!(
        "https://maps.googleapis.com/maps/api/geocode/json?address={}&components=country:MY&region=my&language=ms&key={}",
        urlencoding::encode(&query),
        urlencoding::encode(&key)
    );
    let resp: serde_json::Value = client.get(&url).send().await.ok()?.json().await.ok()?;
    let loc = &resp["results"][0]["geometry"]["location"];
    let lat = loc["lat"].as_f64()?;
    let lng = loc["lng"].as_f64()?;
    Some((lat, lng))
}

async fn geocode_nominatim(cfg: &Config, name: &str, city: &str, state: &str) -> Option<(f64, f64)> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(8))
        .build()
        .ok()?;
    let query = format!("{name}, {city}, {state}, Malaysia");
    let url = format!(
        "https://nominatim.openstreetmap.org/search?format=jsonv2&limit=1&countrycodes=my&q={}",
        urlencoding::encode(&query)
    );
    let mut rb = client.get(&url);
    if let Some(email) = &cfg.nominatim_email {
        rb = rb.query(&[("email", email)]);
    }
    let resp: serde_json::Value = rb
        .header("User-Agent", "sedekahje-bot")
        .send()
        .await
        .ok()?
        .json()
        .await
        .ok()?;
    let first = resp.get(0)?.clone();
    let lat = first["lat"].as_str()?.parse().ok()?;
    let lng = first["lon"].as_str()?.parse().ok()?;
    Some((lat, lng))
}

/// Reverse geocode coords to an address (Google, fallback Nominatim).
pub async fn reverse_geocode(cfg: &Config, lat: f64, lng: f64) -> Option<String> {
    if let Some(key) = cfg.google_geocoding_api_key.as_ref() {
        if !key.is_empty() {
            let client = reqwest::Client::new();
            let url = format!(
                "https://maps.googleapis.com/maps/api/geocode/json?latlng={},{}&key={}&components=country:MY&region=my&language=ms",
                lat, lng, urlencoding::encode(key)
            );
            let resp: serde_json::Value = client.get(&url).send().await.ok()?.json().await.ok()?;
            let addr = resp["results"][0]["formatted_address"].as_str()?;
            if !addr.is_empty() {
                return Some(addr.to_string());
            }
        }
    }
    let client = reqwest::Client::new();
    let url = format!(
        "https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat={}&lon={}",
        lat, lng
    );
    let mut rb = client.get(&url);
    if let Some(email) = &cfg.nominatim_email {
        rb = rb.query(&[("email", email)]);
    }
    let resp: serde_json::Value = rb
        .header("User-Agent", "sedekahje-bot")
        .send()
        .await
        .ok()?
        .json()
        .await
        .ok()?;
    resp["display_name"].as_str().map(String::from)
}

/// Public helper to build a link to an approved institution page.
pub fn institution_link(category: &str, slug: &str) -> String {
    format!("{}/{}/{}", app_url(), category, slug)
}

#[allow(dead_code)]
pub fn unused_placeholder() {}
